File: leetcode/1761_Minimum_Degree_of_a_Connected_Trio_in_a_Graph.py
from typing import List
from unittest import TestCase

class Solution1:
    def minTrioDegree(self, n: int, edges: List[List[int]]) -> int:
        adj_matrix = [[False] * n for _ in range(n)]
        edges_cnt = [0] * n
        min_edges_cnt = n * 3
        for u, v in edges:
            u, v = u -1, v - 1
            adj_matrix[u][v] = True
            adj_matrix[v][u] = True
            edges_cnt[u] += 1
            edges_cnt[v] += 1
        for v in range(n):
            for u in range(v + 1, n):
                for w in range(u + 1, n):
                    if adj_matrix[v][u] and adj_matrix[v][w] and adj_matrix[u][w]:
                        cnt = edges_cnt[v] + edges_cnt[u] + edges_cnt[w]
                        if cnt < min_edges_cnt:
                            min_edges_cnt = cnt
        return -1 if min_edges_cnt == n * 3 else min_edges_cnt - 6

class Solution2:
    def minTrioDegree(self, n: int, edges: List[List[int]]) -> int:
        adj_matrix = [[False] * n for _ in range(n)]
        edges_cnt = [0] * n
        min_edges_cnt = n * 3

        for v, u in edges:
            v, u = (v - 1, u - 1) if v < u else (u - 1, v - 1)
            adj_matrix[v][u] = True
            edges_cnt[u] += 1
            edges_cnt[v] += 1

        for v,u in edges:
            v, u = (v - 1, u - 1) if v < u else (u - 1, v - 1)
            for w in range(u + 1, n):
                if w != v and w != u and adj_matrix[v][w] and adj_matrix[u][w]:
                    cnt = edges_cnt[v] + edges_cnt[u] + edges_cnt[w] - 6
                    if cnt < min_edges_cnt:
                        min_edges_cnt = cnt
        return -1 if min_edges_cnt == n * 3 else min_edges_cnt

class MinTrioDegreeTest(TestCase):
    
    def test_1(self) -> None:
        self.assertEqual(3 ,Solution().minTrioDegree(6, [[1,2],[1,3],[3,2],[4,1],[5,2],[3,6]]))
    
    def test_2(self) -> None:
        self.assertEqual(0 , Solution().minTrioDegree(7, [[1,3],[4,1],[4,3],[2,5],[5,6],[6,7],[7,5],[2,6]]))
    

File: leetcode/1761_Minimum_Degree_of_a_Connected_Trio_in_a_Graph.rs
struct Solution {}

impl Solution {
    pub fn min_trio_degree(n: i32, edges: Vec<Vec<i32>>) -> i32 {
        let n = n as usize;
        let mut adj_matrix = vec![vec![false; n]; n];
        let mut edge_cnts = vec![0; n];
        let mut min_edge_cnt = n * 3;
        for edge in edges {
            let (v, u) = (edge[0] as usize - 1, edge[1] as usize - 1);
            adj_matrix[v][u] = true;
            adj_matrix[u][v] = true;
            edge_cnts[v] += 1;
            edge_cnts[u] += 1;
        }
        for v in 0..n {
            for u in v + 1..n {
                for w in u + 1..n {
                    if adj_matrix[v][u] && adj_matrix[u][w] && adj_matrix[v][w] {
                        let edge_cnt = edge_cnts[v] + edge_cnts[u] + edge_cnts[w];
                        if edge_cnt < min_edge_cnt {
                            min_edge_cnt = edge_cnt;
                        }
                    }
                }
            }
        }
        if min_edge_cnt < n * 3 { min_edge_cnt as i32 - 6 } else { -1 }
    }
}

#[test]
fn test_1() {
    assert_eq!(3 , Solution::min_trio_degree(6, vec![vec![1,2], vec![1,3], vec![3,2], vec![4,1], vec![5,2], vec![3,6]]));
}

#[test]
fn test_2() {
    assert_eq!(0 , Solution::min_trio_degree(7, vec![vec![1,3], vec![4,1], vec![4,3], vec![2,5], vec![5,6], vec![6,7], vec![7,5], vec![2,6]]));
}

